
% Operators
\def\ind#1{\mathbb{I}\left(#1\right)}
\def\evalat#1#2{\left.#1\right|_{#2}}
\def\fracat#1#2#3{\left.\frac{#1}{#2}\right\vert_{#3}}
\def\iid{\overset{iid}{\sim}}
\def\expect#1#2{\underset{#1}{\mathbb{E}}\left[#2\right]}
\def\expecthat#1#2{\underset{#1}{\widehat{\mathbb{E}}}\left[#2\right]}
\def\abs#1{\left|#1\right|}
\def\norm#1{\left\Vert#1\right\Vert}
\def\norminf#1{\left\Vert#1\right\Vert_{\infty}}
\def\normop#1{\left\Vert#1\right\Vert_{\mathrm{op}}}
\def\sumkm{\sum_{\k=1}^{\kmax - 1}}
\def\linop{\mathcal{L}} % A linear operator

\DeclareMathOperator*{\argmax}{\mathrm{argmax}}
\DeclareMathOperator*{\argmin}{\mathrm{argmin}}
\DeclareMathOperator*{\esssup}{\mathrm{esssup}}
\DeclareMathOperator*{\essinf}{\mathrm{essinf}}
\DeclareMathOperator*{\argsup}{\mathrm{argsup}}
\DeclareMathOperator*{\arginf}{\mathrm{arginf}}

% Variables
\def\etaopt{\hat\eta} % Optimal vb parameters
\def\x{x}   % Data
\def\t{t}   % Generic prior parameter
\def\tp{t_{p}}   % Prior parameter for Lp perturbation
\def\z{z}   % Cluster indicators
\def\g{g}   % Function of interest
\def\k{k}   % Cluster index
\def\n{n}   % Data index
\def\b{b}   % additive shift for mice model
\def\l{l}   % index for locus
\def\i{i}   % index for chromosome
\def\d{d}   % data dimension for GMM
\def\s{s}   % sigmoid function
\def\nuk{\nu_{\k}}   % K-th stick.  Have to type this a lot.
\def\const{C}   % Constant
\def\lnu{\tilde{\nu}}   % Unconstrained stick
\def\lnuk{\tilde{\nu}_{\k}}   % Unconstrained stick
\def\lnumean{\eta^{\mu}}   % Unconstrained stick vb mean
\def\lnusd{\eta^{\sigma}}   % Unconstrained stick vb std
\def\hess#1{H_{#1}}   % Hessian
\def\crosshessian{\hat J} % the cross hessian
\def\infl{\Psi}   % The influence function

% Variational parameters
\def\etabeta{\eta_{\beta}}  % VB parameters for certain components
\def\etanu{\eta_{\nu}}  % VB parameters for certain components
\def\etalnu{\eta_{\lnu}}  % VB parameters for certain components
\def\etanuk{\eta_{\nuk}}  % VB parameters for certain components
\def\etaz{\eta_{\z}}  % VB parameters for certain components
\def\etaglob{\eta_{\gamma}}  % VB parameters for theta and nu
\def\etalocal{\eta_{\ell}}  % VB parameters for local variables
\def\etaoptglob{\etaopt_{\gamma}}  % optimal VB parameters for theta and nu
\def\etaoptlocal{\etaopt_{\ell}}  % optimal VB parameters for local variables
\def\etaopttheta{\etaopt_{\theta}}  % VB parameters for certain components
\def\etaoptnu{\etaopt_{\nu}}  % VB parameters for certain components
\def\etaoptnuk{\etaopt_{\nuk}}  % VB parameters for certain components
\def\etaoptz{\etaopt_{\z}}  % VB parameters for certain components
\def\etaoptgamma{\etaopt_{\gamma}}  % VB parameters for certain components
\def\hessopt{\hat{H}}  % VB parameters for certain components


% Distributions central to the story
\def\p{\mathcal{P}}   % Other distributions
\def\q{\mathcal{Q}}   % VB dist
\def\ptil{\tilde{\p}}   % Other distributions
\def\qtil{\tilde{\q}}   % Unnormalized VB dist
\def\pstick{\p_{\mathrm{stick}}} % Stick breaking dens, use only for BNP section
\def\pbetaprior{\p_{\mathrm{base}}}   % distribution on cluster centroids
\def\lqgrad#1{{\nabla_\eta \log \qtil}\left(#1\right)}   % Log VB distribution gradient
\def\lqgradbar#1{\overline{\lqgrad{#1}}\,\,}   % Log VB distribution gradient centered
\def\pbase{\p_{0}}   % Initial (base) prior (different from the base prior of the DP)
\def\palt{\p_{1}}   % Alternative prior
\def\pbasetil{\ptil_{0}}   % Base prior
\def\palttil{\ptil_{1}}   % Alternative prior

% Less commonly used distributions
\def\normdist#1{\mathcal{N}\left(#1\right)}   % Normal distribution
\def\gem{\mathrm{GEM}(\alpha)}
\def\KL#1{\mathrm{KL}\left(#1\right)}   % KL divergence
\def\normalwishart#1{\mathcal{N}\mathcal{W}\left(#1\right)}   % Wishart distribution
\def\gammadist#1{\mathrm{Gamma}\left(#1\right)}   % Gamma distribution
\def\betadist#1{\mathrm{Beta}\left(#1\right)}   % Gamma distribution
\def\KLglobal{\mathrm{KL}_{\mathrm{glob}}}   % KL divergence as function of global parameters only

% Taylor series
\def\etalin{\etaopt^{\mathrm{lin}}}
\def\etalinglobal{\etaopt_\gamma^{\mathrm{lin}}}
\def\glin{\g^{\mathrm{lin}}}


% Dimensions
\def\N{N}   % Number of datapoints
\def\K{K}   % Number of components
\def\kmax{{\K_{\mathrm{max}}}}   % Truncation
\def\etadim{{D_{\eta}}}     % VB parameter
\def\betadim{{D_{\beta}}}   % Centroid parameter
\def\zetadim{{D_{\zeta}}}   % All parameters
\def\ngh{N_{\mathrm{GH}}}   % Number of GH points
\def\thetadim{{D_{\theta}}} % A generic parameter

% Domains
\def\etadom{\Omega_{\eta}}
\def\thetadom{\Omega_{\theta}}
\def\betadom{\Omega_{\beta}}
\def\tdom{\Omega_{\t}}
\def\linf{{L_{\infty}}}
\def\ball{\mathcal{B}}
\def\ballclosed{\overline{\ball}}

% Annotations
\def\mathtxt#1{\quad\textrm{#1}\quad}%
\def\mathand{\quad\textrm{and}\quad}%
\def\mathwhere{\quad\textrm{where}\quad}%
\def\constdesc#1{\textrm{(}\const\textrm{ does not depend on }#1\textrm{)}}
\def\assuitemref#1#2{\assuref{#1} (\itemref{#2})}%

% some posterior statistics
\newcommand{\gclusters}{\g_{\text{cl},\tau}} % number of in-sample clusters (thresholded)
\newcommand{\gclusterszero}{\g_{\text{cl},0}} % number of in-sample clusters (not thresholded)
\newcommand{\gclustersabbr}{\g_{\text{cl}}} % number of in-sample clusters (not thresholded), abbreviated
\newcommand{\gclusterspred}{\g_{\text{pred,cl},\tau}} % number of predictive clusters
\newcommand{\gclusterspredzero}{\g_{\text{pred,cl},0}} % number of predictive clusters
\newcommand{\gclusterspredabbr}{\g_{\text{pred,cl}}} % number of predictive clusters
\newcommand{\gcoclustering}{\g_{\text{cc}}} % coclustering matrix
\newcommand{\laplacianevsum}{\g_{\text{ev}}} % the sum of the eigenvalues of the graph-laplacian (for mice)
\newcommand{\gpop}{\g_{\text{pop}}} % number of in-sample populations (for structure)
\newcommand{\gloci}{\g_{\text{loci}}} %  number of loci in a population(for structure)
\newcommand{\gadmix}{\g_{\text{admix}}} % expected admixture (for structure)
\newcommand{\nclusters}{G_{\text{cl}}} % random variable for number of clusters

% commands for for mice data
\newcommand{\ngenes}{N} % number of genes
\newcommand{\ntimepoints}{M} % number of timepoints
\newcommand{\regmatrix}{A}
\newcommand{\timeindx}{T} % would just use \t, but then this parameterizes a distribution ...
\newcommand{\pshift}{\p_{\mathrm{shift}}}

% commands for for structure data
\newcommand{\nindiv}{N} % number of individuals genotyped
\newcommand{\nloci}{L} % number of loci
\newcommand{\latentpop}{\beta} % latent population frequencies
\newcommand{\latentadmix}{\pi} % latent admixtures
\def\categoricaldist#1{\mathrm{Categorical}\left(#1\right)}
\newcommand{\phiworstcase}{\phi_{\mathrm{wc}}} % worst-case perturbation
